File: tests/test_diff.py
import numpy as np
import pytest

from yuvdiff.diff import DiffEngine, DiffResult
from yuvdiff.formats import BitDepth, PixelFormat
from yuvdiff.parser import YUVFrame


def make_frame(y, u=None, v=None, w=4, h=4, bit_depth=8, fmt=PixelFormat.YUV420P):
    if u is None:
        u = np.zeros((h // 2, w // 2), dtype=np.uint8)
    if v is None:
        v = np.zeros((h // 2, w // 2), dtype=np.uint8)
    return YUVFrame(
        y=np.asarray(y, dtype=np.uint8),
        u=np.asarray(u, dtype=np.uint8),
        v=np.asarray(v, dtype=np.uint8),
        bit_depth=bit_depth,
        width=w,
        height=h,
        pixel_format=fmt,
    )


class TestDiffBasic:
    def test_identical_frames_zero_diff(self):
        y = np.full((4, 4), 100, dtype=np.uint8)
        u = np.full((2, 2), 50, dtype=np.uint8)
        v = np.full((2, 2), 50, dtype=np.uint8)
        a = make_frame(y, u, v)
        b = make_frame(y, u, v)
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert isinstance(result, DiffResult)
        assert result.diff_y.sum() == 0
        assert result.diff_u.sum() == 0
        assert result.diff_v.sum() == 0
        assert result.mask.sum() == 0
        assert result.diff_pixel_count == 0
        assert result.total_pixel_count == 16

    def test_black_vs_white_full_diff(self):
        a = make_frame(np.zeros((4, 4), dtype=np.uint8))
        b = make_frame(np.full((4, 4), 255, dtype=np.uint8))
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert result.diff_y.max() == 255
        assert result.mask.all()
        assert result.diff_pixel_count == 16

    def test_threshold_boundary_excludes_equal(self):
        y_a = np.zeros((4, 4), dtype=np.uint8)
        y_b = np.full((4, 4), 4, dtype=np.uint8)  # diff = 4 == threshold
        a = make_frame(y_a)
        b = make_frame(y_b)
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert result.diff_y.max() == 4
        assert result.mask.sum() == 0

    def test_threshold_boundary_includes_above(self):
        y_a = np.zeros((4, 4), dtype=np.uint8)
        y_b = np.full((4, 4), 5, dtype=np.uint8)  # diff = 5 > threshold
        a = make_frame(y_a)
        b = make_frame(y_b)
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert result.mask.sum() == 16


class TestDiffPerChannel:
    def test_only_y_differs(self):
        y_a = np.zeros((4, 4), dtype=np.uint8)
        y_b = np.full((4, 4), 100, dtype=np.uint8)
        u = np.full((2, 2), 50, dtype=np.uint8)
        v = np.full((2, 2), 50, dtype=np.uint8)
        a = make_frame(y_a, u, v)
        b = make_frame(y_b, u, v)
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert result.diff_u.sum() == 0
        assert result.diff_v.sum() == 0
        assert result.diff_y.max() == 100
        assert result.mask.sum() == 16

    def test_chroma_diff_upsampled_to_y_resolution(self):
        y = np.zeros((4, 4), dtype=np.uint8)
        u_a = np.zeros((2, 2), dtype=np.uint8)
        u_b = np.array([[100, 0], [0, 0]], dtype=np.uint8)
        v = np.zeros((2, 2), dtype=np.uint8)
        a = make_frame(y, u_a, v)
        b = make_frame(y, u_b, v)
        engine = DiffEngine(threshold=4)
        result = engine.diff(a, b)
        assert result.mask[:2, :2].all()
        assert not result.mask[:2, 2:].any()
        assert not result.mask[2:, :].any()
        assert result.diff_pixel_count == 4


class TestDiffShape:
    def test_mask_shape_matches_y(self):
        y = np.zeros((4, 4), dtype=np.uint8)
        a = make_frame(y)
        b = make_frame(y + 1)
        engine = DiffEngine(threshold=0)
        result = engine.diff(a, b)
        assert result.mask.shape == (4, 4)

File: cpp/tests/test_diff.cpp
#include "test_framework.hpp"
#include "yuvdiff/diff.hpp"

using namespace yuvdiff;

namespace {

YUVFrame make_frame(int w, int h, uint8_t y_val, uint8_t u_val, uint8_t v_val, PixelFormat fmt = PixelFormat::YUV420P) {
    YUVFrame f;
    f.width = w;
    f.height = h;
    f.bit_depth = 8;
    f.pixel_format = fmt;

    auto [hf, vf] = chroma_subsampling(fmt);
    int cw = w / hf;
    int ch = h / vf;

    f.y8.resize(w * h, y_val);
    f.u8.resize(cw * ch, u_val);
    f.v8.resize(cw * ch, v_val);
    return f;
}

} // anonymous namespace

TEST_CASE(test_diff_identical_frames) {
    YUVFrame a = make_frame(16, 16, 100, 128, 128);
    YUVFrame b = make_frame(16, 16, 100, 128, 128);

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);

    ASSERT_EQ(res.diff_pixel_count, 0LL);
    ASSERT_EQ(res.total_pixel_count, 256LL);
    for (int v : res.diff_y) ASSERT_EQ(v, 0);
    for (int v : res.diff_u) ASSERT_EQ(v, 0);
    for (int v : res.diff_v) ASSERT_EQ(v, 0);
    for (uint8_t m : res.mask) ASSERT_EQ(m, 0);
}

TEST_CASE(test_diff_black_vs_white) {
    YUVFrame a = make_frame(16, 16, 0, 0, 0);
    YUVFrame b = make_frame(16, 16, 255, 255, 255);

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);

    ASSERT_EQ(res.diff_pixel_count, 256LL);
    for (int v : res.diff_y) ASSERT_EQ(v, 255);
    for (uint8_t m : res.mask) ASSERT_EQ(m, 1);
}

TEST_CASE(test_diff_threshold_boundary) {
    YUVFrame a = make_frame(16, 16, 100, 128, 128);
    YUVFrame b = make_frame(16, 16, 104, 128, 128); // diff == 4

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);
    ASSERT_EQ(res.diff_pixel_count, 256LL); // diff == 4 > 0
    ASSERT_EQ(res.diff_gt_2t, 0LL);        // diff == 4 is not > 8 (2t)
    ASSERT_EQ(res.diff_gt_t, 0LL);         // diff == 4 is not > 4 (t)
    ASSERT_EQ(res.diff_gt_half_t, 256LL);  // diff == 4 > 2 (t/2)

    YUVFrame c = make_frame(16, 16, 105, 128, 128); // diff == 5 > 4
    DiffResult res2 = engine.diff(a, c);
    ASSERT_EQ(res2.diff_pixel_count, 256LL);
    ASSERT_EQ(res2.diff_gt_t, 256LL);
    ASSERT_EQ(res2.diff_gt_2t, 0LL);       // diff == 5 is not > 8 (2t)

    YUVFrame d = make_frame(16, 16, 109, 128, 128); // diff == 9 > 8 (2t)
    DiffResult res3 = engine.diff(a, d);
    ASSERT_EQ(res3.diff_gt_2t, 256LL);
}

TEST_CASE(test_diff_chroma_upsampling) {
    YUVFrame a = make_frame(16, 16, 100, 128, 128, PixelFormat::YUV420P);
    YUVFrame b = make_frame(16, 16, 100, 128, 128, PixelFormat::YUV420P);

    // Modify only one chroma pixel in U
    b.u8[0] = 200; // diff = 72 > 4

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);

    // In 420P, one chroma pixel corresponds to a 2x2 block in Y resolution = 4 pixels
    ASSERT_EQ(res.diff_pixel_count, 4LL);
    ASSERT_EQ(res.diff_gt_t, 4LL);
    ASSERT_EQ(res.diff_gt_2t, 4LL);
}

TEST_CASE(test_diff_cross_depth) {
    YUVFrame a = make_frame(16, 16, 100, 128, 128); // 8-bit Y=100 -> equivalent 10-bit Y=400
    YUVFrame b;
    b.width = 16;
    b.height = 16;
    b.bit_depth = 10;
    b.pixel_format = PixelFormat::YUV420P;
    b.y16.resize(16 * 16, 400); // exactly matching
    b.u16.resize(8 * 8, 512);   // 128 * 4 = 512
    b.v16.resize(8 * 8, 512);

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);
    ASSERT_EQ(res.diff_pixel_count, 0LL);
}

TEST_CASE(test_diff_cross_chroma_format) {
    YUVFrame a = make_frame(16, 16, 100, 128, 128, PixelFormat::YUV420P);
    YUVFrame b = make_frame(16, 16, 100, 128, 128, PixelFormat::YUV422P);

    DiffEngine engine(4);
    DiffResult res = engine.diff(a, b);
    ASSERT_EQ(res.diff_pixel_count, 0LL);
}

TEST_CASE(test_diff_statistics) {
    YUVFrame a = make_frame(4, 4, 10, 128, 128);
    YUVFrame b = make_frame(4, 4, 10, 128, 128);

    // Set specific diffs: 0, 2, 4, 10 across 16 pixels
    // 4 pixels with diff 0 (10 vs 10)
    // 4 pixels with diff 2 (10 vs 12)
    // 4 pixels with diff 4 (10 vs 14)
    // 4 pixels with diff 10 (10 vs 20)
    for (int i = 4; i < 8; ++i) b.y8[i] = 12;
    for (int i = 8; i < 12; ++i) b.y8[i] = 14;
    for (int i = 12; i < 16; ++i) b.y8[i] = 20;

    DiffEngine engine(4); // t=4, t/2=2, 2t=8
    DiffResult res = engine.diff(a, b);

    // Non-zero diff pixels (d > 0): 12 pixels (values: 4x2, 4x4, 4x10)
    ASSERT_EQ(res.diff_pixel_count, 12LL);
    ASSERT_EQ(res.total_pixel_count, 16LL);
    ASSERT_NEAR(res.diff_ratio, 12.0 / 16.0, 0.0001);

    // d > 8 (2t): only the 4 pixels with diff 10
    ASSERT_EQ(res.diff_gt_2t, 4LL);
    ASSERT_NEAR(res.diff_gt_2t_ratio, 4.0 / 16.0, 0.0001);

    // d > 4 (t): only the 4 pixels with diff 10
    ASSERT_EQ(res.diff_gt_t, 4LL);
    ASSERT_NEAR(res.diff_gt_t_ratio, 4.0 / 16.0, 0.0001);

    // d > 2 (t/2): the 4 pixels with diff 4 and 4 pixels with diff 10 = 8 pixels
    ASSERT_EQ(res.diff_gt_half_t, 8LL);
    ASSERT_NEAR(res.diff_gt_half_t_ratio, 8.0 / 16.0, 0.0001);

    // Stats over d > 0 pixels:
    ASSERT_EQ(res.diff_min, 2);
    ASSERT_EQ(res.diff_max, 10);
    // Mean = (2*4 + 4*4 + 10*4) / 12 = (8 + 16 + 40) / 12 = 64 / 12 = 5.3333...
    ASSERT_NEAR(res.diff_mean, 64.0 / 12.0, 0.001);
    // Diffs sorted: [2,2,2,2, 4,4,4,4, 10,10,10,10] -> median = 4.0
    ASSERT_NEAR(res.diff_median, 4.0, 0.001);
}
